### PR TITLE
ARROW-5490: [C++] Remove ARROW_BOOST_HEADER_ONLY

### PR DESCRIPTION
**Pull request by pitrou**
That CMake variable isn't exposed as an option and probably doesn't work anymore.

### CODE REVIEW

File: cpp/src/arrow/CMakeLists.txt [hunk 1]
```diff
@@ -126,6 +126,8 @@ set(ARROW_SRCS
     io/buffered.cc
     io/compressed.cc
     io/file.cc
+    io/hdfs.cc
```
Comment (fsaintjacques) [inline cpp/src/arrow/CMakeLists.txt]: Should this be conditional on `ARROW_HDFS`?

Comment (pitrou) [reply cpp/src/arrow/CMakeLists.txt]: I don't know. @wesm What do you think?

Comment (wesm) [reply cpp/src/arrow/CMakeLists.txt]: There are no additional build-time dependencies to compile this, so I think it's OK to have it here

**Comment by wesm** [comment]
cc @majetideepak -- there are diminishing returns to maintaining this unless we add a "No Boost" docker-compose build.

**Comment by majetideepak** [comment]
+1

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**